### DISCUSSION TITLE
Is there any kind of proxy support?

### DISCUSSION THREAD

**Discussion by YetAnotherGerrit** [ignacio-chiazzo/ruby_whatsapp_sdk#161] [General]
Is there a way to configure a proxy server through that the gem should contact the Meta API?

**Comment by ignacio-chiazzo** [comment]
I'm not aware of a proxy using Cloud API.

**Reply by YetAnotherGerrit** [reply]
I mean, in case my application needs a proxy to communicate with "the internet", therefore also with the Meta API, can this be configured to be used by the gem?

P.S.: I just learned that `Net::HTTP` uses the env-Variable `HTTP_PROXY` if found. I will try that and circle back.

**Reply by ignacio-chiazzo** [reply]
Not sure if I follow. I'm curious to see what you find. Please share later, thanks!

**Reply by YetAnotherGerrit** [reply]
Okay, let me enhance this with more information. Also, we did some testing.

Use-Case: I have an application using your gem on Server A. This server is in a network with no direct internet connection. There is a HTTP proxy in the network on Server B that provides access to the internet.

Finding: when setting the ENV variable HTTP_PROXY to http://server-b.local the gem can reach the WhatsApp.

Enhancement: provide the address to the proxy (Server B in above use case) not only by ENV variable, but also optional during initialization of the gem.

**Reply by ignacio-chiazzo** [reply]
oh I see. So instead of hitting `https://graph.facebook.com/<api_resource>` from server-a you want to hit `server-b.local/<api_resource>` and your server-b would redirect the request to `https://graph.facebook.com/<api_resource>`. 

If so, there are some small changes required in the gem: 
1- Add the ability to set in the initializer a `HTTP_PROXY`. 
2- Update [API configuration](https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/89a999af6c206834ca84f12a1fa0c8d82fd54de6/lib/whatsapp_sdk/api/api_configuration.rb#L7) to use the HTTP_PROXY or the API_URL


The URL is only [used here ](https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/89a999af6c206834ca84f12a1fa0c8d82fd54de6/lib/whatsapp_sdk/api/client.rb#L51C30-L51C46).

**Reply by YetAnotherGerrit** [reply]
Actually, you are still hitting `https://graph.facebook.com/<api_resource>` but alongside the request you provide credentials to use `server-b.local` as a proxy.

So when you issue a Net::HTTP call here:
https://github.com/ignacio-chiazzo/ruby_whatsapp_sdk/blob/89a999af6c206834ca84f12a1fa0c8d82fd54de6/lib/whatsapp_sdk/api/client.rb#L68-L75

You support proxy parameters like described here, if provided optionally on initialization of your gem:
https://github.com/ruby/net-http/blob/71f8f24eb9db7973b0cddda4778c308baa538101/lib/net/http.rb#L348-L371

Thanks a ton, for looking into it! <3
